[ao][sparsity] Training-aware data sparsity callback for lightning

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81813
* __->__ #80371
* #80370

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestTrainingAwareCallback```

Commit:
[ao][sparsity] Training-aware data sparsity callback for lightning

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Training-aware data sparsity callback for lightning"


This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Training-aware data sparsity callback for lightning"


This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py TestTrainingAwareCallback```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Training-aware data sparsity callback for lightning"


This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py TestTrainingAwareCallback```

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:
```diff
@@ -1,14 +1,18 @@
         if type(parameter) in SUPPORTED_TYPES:
             valid_name = _get_valid_name(name)
-            data_sparsifier.add_data(name=valid_name, data=parameter)  # will be defaulted to default configs
+            if config is None:
+                config = {}
```
Can we take that out of the for loop?

macandro96 replied:
Oops - done.

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:
```diff
@@ -1,18 +1,27 @@
 import torch
 from dataclasses import dataclass, field
 from typing import Any, Optional, Dict
 import pytorch_lightning as pl
```
```
import pytorch_lightning as pl  # type: ignore
```

macandro96 replied:
This still gives me TypeIgnore error - ```This line has an unqualified `type: ignore`; please convert it to `type:
    ignore[xxxx]```. I couldn't find like a specific code for this.

macandro96 replied:
```type: ignore[import]``` seemed to fix this.

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:
```diff
@@ -1,18 +1,27 @@
+)
 
 
 @dataclass
```
This is a little tricky: You are making it a dataclass, and you are inheriting from a callback -- this makes me concerned about the MRO. Also, AFAIK, the dataclass overrides the `__annotations__`, `__slots__`, and `__hash__`. imho, this might cause some problems if the Callbacks decide to use those.

macandro96 replied:
Yeah, this might be concerning. I've removed the data class and used a simple class with __init__() for now.

z-a-f reviewed (changes requested):
OVerall great, minor comments

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -78,5 +105,121 @@ def test_post_training_callback(self):
+            self._check_on_train_epoch_end(pl_module, callback)
+
+
+class TestTrainingAwareCallback(_TrainingAwareCallbackTestCase):
```
nit: should it be SparsityAwareTrainingCallback?

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -78,5 +105,121 @@ def test_post_training_callback(self):
+
+        data_sparsifier_state_dict = callback.data_sparsifier.state_dict()
+
+        # compare container objects
```
nit: can we give a bit more detail on these comparisons? e.g. why do we compare them, what is the expected result

macandro96 replied:
done

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -78,5 +105,121 @@ def test_post_training_callback(self):
+    """
+    def _check_on_train_start(self, pl_module, callback, sparsifier_args, scheduler_args):
+
+        callback.on_train_start(42, pl_module)  # 42 is a dummy value
```
42 --> should we provide a DummyTrainer to match the expected type?

macandro96 replied:
Adding DummyTrainer seemed to raise some backend cuda error. Since, we are testing each function, it should be okay for now I think.

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -78,5 +105,121 @@ def test_post_training_callback(self):
+    Simulates model training and makes sure that each hook is doing what is expected
+    """
+    def _check_on_train_start(self, pl_module, callback, sparsifier_args, scheduler_args):
+
```
would also be helpful if we can add some comments for these individual checks, describing in a high level what do we do here and what is the expected result

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:
```diff
@@ -26,3 +30,8 @@ def _log_sparsified_level(model, data_sparsifier) -> None:
             f"Sparsity in layer {name} = {sparsity_level: .2%}"
         )
+
+def _create_data_scheduler(data_sparsifier, data_scheduler_args, data_scheduler_type):
```
looks like this function is only used once, should we just inline the code?

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:
```diff
@@ -1,14 +1,18 @@
 
-def _attach_model_to_data_sparsifier(module, data_sparsifier):
+def _attach_model_to_data_sparsifier(module, data_sparsifier, config=None):
     for name, parameter in module.named_parameters():
```
can we add some comments for this function?

macandro96 replied:
done

Commit:
Update on "[ao][sparsity] Training-aware data sparsity callback for lightning"


This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py TestTrainingAwareCallback```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Training-aware data sparsity callback for lightning"


This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestTrainingAwareCallback```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Training-aware data sparsity callback for lightning"


This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestTrainingAwareCallback```

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -90,5 +118,166 @@ def test_post_training_callback(self):
         self.run_all_checks(pl_module, callback, sparsifier_args)
 
 
+class _TrainingAwareCallbackTestCase(TestCase):
```
I smell the abuse of OOP :)

macandro96 replied:
Merged the test classes

Commit:
Update on "[ao][sparsity] Training-aware data sparsity callback for lightning"


This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestTrainingAwareCallback```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Training-aware data sparsity callback for lightning"


This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using
<callback_obj>.sparsified

Test Plan:
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestTrainingAwareCallback```

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g
